KEP-1880 alpha: IPAddresses (#37620)
kubernetes/website: content/en/docs/reference/networking/virtual-ips.md

appropriate endpoint. The environment variables and DNS for Services are actually
populated in terms of the Service's virtual IP address (and port).

> 

Reviewer: ```suggestion

### IP address allocation tracking

```
Maybe use [tabs](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#tabs) to describe and compare the two options (not sure what we call them, but basically the old way with a global allocation map in etcd that end users can't see, vs. the new alpha way).

We usually put the feature state shortcode under a heading that summarises the thing. _Not_ the change, but the thing for which the behavior is being changed.

Author: there is no much visible change except the new objects
```
 kubectl get services
NAME         TYPE        CLUSTER-IP        EXTERNAL-IP   PORT(S)   AGE
kubernetes   ClusterIP   2001:db8:1:2::1   <none>        443/TCP   3d1h
aojea@aojea:~/src/kind$ kubectl get ipaddresses
NAME              PARENTREF
2001:db8:1:2::1   services/default/kubernetes
2001:db8:1:2::a   services/kube-system/kube-dns
```

do you want me to add it?

{{< feature-state for_k8s_version="v1.27" state="alpha" >}}
If you enable the `MultiCIDRServiceAllocator`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/),
the `ClusterIP` address associated to each `Service` will have a referenced

> `IPAddress` object.

Reviewer: (nit) avoid writing API kinds within backticks

Author: done

assignments (e.g. due to administrator intervention) and for cleaning up allocated
IP addresses that are no longer used by any Services.

{{< feature-state for_k8s_version="v1.27" state="alpha" >}}
If you enable the `MultiCIDRServiceAllocator`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/),
the control plane replaces the existing etcd allocator with a new one, using IPAddress
objects instead of an internal global allocation map.  The ClusterIP address

> associated to each `Service` will have a referenced IPAddress object.

Reviewer: Do you also need to enable an alpha API group for this to actually work?

Author: yeah, good catch, it needs to enable networking.k8s.io/v1alpha1

> The background allocator is also replaced by a new one to handle the new IPAddress

Author: ```suggestion
The background controller is also replaced by a new one to handle the new IPAddress
```

objects and the migration from the old allocator model.

One of the main benefits of the new allocator is that it removes the size limitations
for the `service-cluster-ip-range`, there is no limitations for IPv4 and for IPv6
users can use masks equal or larger than /64 (previously it was /108).

Users now will be able to inspect the IP addresses assigned to their Services, and
new network APIs, like Gateway API, can use this new object to extend the Kubernetes
networking capabilities overcoming the limitations of current Services API.

> ```bash
> $ kubectl get services
> NAME         TYPE        CLUSTER-IP        EXTERNAL-IP   PORT(S)   AGE
> kubernetes   ClusterIP   2001:db8:1:2::1   <none>        443/TCP   3d1h
> 
> $ kubectl get ipaddresses
> NAME              PARENTREF
> 2001:db8:1:2::1   services/default/kubernetes
> 2001:db8:1:2::a   services/kube-system/kube-dns
> ```

Reviewer 2: ````suggestion
```shell
kubectl get services
```
```
NAME         TYPE        CLUSTER-IP        EXTERNAL-IP   PORT(S)   AGE
kubernetes   ClusterIP   2001:db8:1:2::1   <none>        443/TCP   3d1h
```
```shell
kubectl get ipaddresses
```
```
NAME              PARENTREF
2001:db8:1:2::1   services/default/kubernetes
2001:db8:1:2::a   services/kube-system/kube-dns
```
````

assignments (e.g. due to administrator intervention) and for cleaning up allocated
IP addresses that are no longer used by any Services.

{{< feature-state for_k8s_version="v1.27" state="alpha" >}}
If you enable the `MultiCIDRServiceAllocator`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/),
the control plane replaces the existing etcd allocator with a new one, using IPAddress
objects instead of an internal global allocation map.  The ClusterIP address
associated to each `Service` will have a referenced IPAddress object.

The background controller is also replaced by a new one to handle the new IPAddress
objects and the migration from the old allocator model.

One of the main benefits of the new allocator is that it removes the size limitations
for the `service-cluster-ip-range`, there is no limitations for IPv4 and for IPv6
users can use masks equal or larger than /64 (previously it was /108).

Users now will be able to inspect the IP addresses assigned to their Services, and

> new network APIs, like Gateway API, can use this new object to extend the Kubernetes
> networking capabilities overcoming the limitations of current Services API.

Reviewer: ```suggestion
Kubernetes extensions such as the [Gateway](https://gateway-api.sigs.k8s.io/) API, can use this new
IPAddress object kind to enhance the Kubernetes networking capabilities, going beyond the limitations of
the built-in Service API.
```
Optionally, add
```markdown

For example, an extension can monitor IP address allocations and reconcile these with external tooling
such as an allow list manager.
```

{{< feature-state for_k8s_version="v1.27" state="alpha" >}}
If you enable the `MultiCIDRServiceAllocator`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) and the

> [networking.k8s.io/v1alpha1 API group](https://kubernetes.io/docs/tasks/administer-cluster/enable-disable-api/),

Reviewer: ```suggestion
[`networking.k8s.io/v1alpha1` API group](/docs/tasks/administer-cluster/enable-disable-api/),
```

the control plane replaces the existing etcd allocator with a new one, using IPAddress
objects instead of an internal global allocation map.  The ClusterIP address

> associated to each `Service` will have a referenced IPAddress object.

Reviewer: ```suggestion
associated to each Service will have a referenced IPAddress object.
```
